File: pset1/credit/credit.c
# include <stdio.h>
# include <unistd.h>
# include <cs50.h>

int main()
{
    // Declare Number Digit and Sum of Total
    int num_digit = 0, total = 0;
 
    // Declare Variables to find First Two Digits
    int recent = 0, current = 0;
    
    // Request User's Input
    long credit = 0;
    do
    {
        credit = get_long("Number: ");
    }
    while (credit <= 0);
    
    // Calculating Credit Number from Last Digit
    while (credit > 0)
    {
        recent = current;
        
        // Take Current Digit from Last Digit
        current = credit % 10;
        
        if (num_digit % 2 == 0)
        {
            total += current;
        }
        
        else
        {
            int mul = current * 2;
            total += (mul % 10) + (mul / 10);
        }
        
        num_digit++;
        credit /= 10;        
    }
    
    
    if (total % 10 == 0)
    {
        int two_digits = (current * 10) + recent;
        
        // Americal Express
        if (num_digit == 15 && (two_digits == 34 || two_digits == 37))
        {
            printf("AMEX\n");
        }
        
        // Master Card
        else if (num_digit == 16 && (two_digits >= 51 && two_digits <= 55))
        {
            printf("MASTERCARD\n");
        }
        
        // VISA
        else if ((num_digit >= 13 && num_digit <= 16) && current == 4)
        {
            printf("VISA\n");
        }
        
        else
        {
            printf("INVALID\n");
        }
    }
    
    else 
    {
        printf("INVALID\n");
    }

}


    
    

File: pset6/credit/credit.py
import re
from cs50 import get_string


def main():

    # Get card number from user input (stdin)
    cardNumber = None
    while cardNumber == None:
        cardNumber = get_string("Number: ")

    # Check the card Number
    if (luhn_checksum(cardNumber) != 0):
        print("INVALID")
        return 1

    # Match the card pattern
    print(match(cardNumber))


# Generate Luhn's checksum by using slice notation [start:end:step]
# Reference https://railsware.com/blog/python-for-machine-learning-indexing-and-slicing-for-lists-tuples-strings-and-other-sequential-types/
def luhn_checksum(cardNumber):

    total = 0

    # Starting from the last index of the array, add number
    for i, j in cardNumber[-1:0]:

        # Sum the digits which won't multiple by 2 starting from last digit
        total += int(i)

        # Multiply every other digit by 2, starting with the number’s second-to-last digit
        j = int(j) * 2
        total += (j % 10) + (j // 10)

    return total % 10


# Match the card pattern using Regex Expression (re)
def match(cardNumber):

    # Americal Express
    if re.match(r"^3[4-7]\d{13}", cardNumber):
        return "AMEX"

    # MasterCard
    elif re.match(r"^(22)|(5[1-5])\d{14}", cardNumber):
        return "MASTERCARD"

    # VISA
    elif re.match(r"^4\d{12,15}", cardNumber):
        return "VISA"

    else:
        return "INVALID"


if __name__ == "__main__":
    main()